tutorials/rag/agentic_rag.py

```diff
@@ -0,0 +1,274 @@
+"""
+This script demonstrates how to build a more agentic RAG pipeline using Humanloop and OpenAI's API.
+
+The model can choose to call the retrieval tool, or instead ask for clarification from the user.
+
+This leverages tool calling: https://platform.openai.com/docs/guides/function-calling
+
+It assumes that Prompts and Tools are managed in the client's code and logging and evals are done via Humanloop.
+"""
+
+import json
+from dotenv import load_dotenv
+import os
+from chromadb import chromadb
+from openai import OpenAI
+import pandas as pd
+from datetime import datetime
+from humanloop import Humanloop
+
+# Load .env file that contains API keys
+load_dotenv()
+
+# Init clients
+hl = Humanloop(api_key=os.getenv("HUMANLOOP_KEY"))
+chroma = chromadb.Client()
+openai = OpenAI(api_key=os.getenv("OPENAI_KEY"))
+
+# Init the knowledge base - use whatever vector DB you have
+print("Populating knowledge base")
+collection = chroma.get_or_create_collection(name="MedQA")
+knowledge_base = pd.read_parquet("../../assets/sources/textbooks.parquet")
+knowledge_base = knowledge_base.sample(10, random_state=42)
+collection.add(
+    documents=knowledge_base["contents"].to_list(),
+    ids=knowledge_base["id"].to_list(),
+)
+print("Knowledge base populated")
+
+APP_NAME = "agentic-rag-demo"
+
+# Define all the parameters for your pipeline
+MODEL = "gpt-4o-mini"
+# Define your Prompt template that contains variables for answer options required by this example
+# and instructions for tool use
```

Comment:
this sentence is quite hard to read. the "for answer options required by this example and instructions for tool use" bit.

Reply:
Tweaked